core/src/main/java/org/springframework/security/access/expression/AbstractSecurityExpressionHandler.java

```diff
@@ -102,22 +106,47 @@ protected abstract SecurityExpressionOperations createSecurityExpressionRoot(
 			Authentication authentication, T invocation);
 
 	protected RoleHierarchy getRoleHierarchy() {
+		if(! roleHierarchySet && context != null) {
```

Comment:
I don't know how much of an improvement it is, but I've made the change you requested and updated this PR accordingly.